README.md
# DevPail - a bucket of development tooling

*This is version **2.1.1**.*

[![GitHub package.json version][shield-gh-json-version]][gh-repo]
[![License][shield-gh-license]][gh-license]
[![Package Version (npm)][shield-npm-version]][npm-package]
[![Downloads Per Month (npm)][shield-npm-downloads-month]][npm-versions]
[![Buy Me A Coffee][shield-bmac]][bmac]

**DevPail** is what you carry all your DevTools in.  <img align="right" alt="DevPail Logo" src="images/devpail-logo-150x205.png">

**DevPail** builds flexible [Docker] container and
[GulpJS] task runner combinations which isolate and
encapsulate your projects' development environments, so you don't have to.

You don't need to write a custom Gulpfile for each project, but instead
configure the operation of Gulp via directives in `package.json` which enable
various plugins.

## Installation

Before you use **DevPail**, you will need a working installation of
[Docker].

Since **DevPail** can (should!) be used across multiple projects, it's best to
install it globally:

```console
npm install --global devpail
```

If you intend modify the code before use:

```console
git clone https://github.com/tdesposito/DevPail.git
cd DevPail
npm link .
```

## Usage

Before you can start using **DevPail**, you'll need to build at least the
default image (python3.8/node14):

```console
$ devpail --build
DevPail: Building image "devpail:default"...
... lots of stuff happens ...
```

You can pass Docker ARG parameters directly on the above to customize the build.

> **WARNING: The Docker image DevPail builds will contain various secrets from your local machine, including AWS access keys, SSH keys, and Git credentials. *DO NOT PUBLISH OR SHARE YOUR DevPail IMAGES*!**

To initialize your project, in the top-level project directory (where your
`package.json` and other project files are):

```console
$ devpail --init
DevPail: Setting defaults for your project...
```

**DevPail** creates a Docker volume to contain all your development tooling
(NodeJS modules, Python packages, and so forth), and a Docker container which
mounts that volume and your source code. Inside that container, it runs a
plugin-based GulpJS task runner. Rather than constructing a bespoke
`gulpfile.js` for each project, you configure **DevPail** plugins via your
`package.json` file. How? Oh, look, that's right here!

## Configuration

Configuration for your **DevPail** is kept under the `devpail` key in your
`package.json` file. You can use `devpail --init` to provide you with a
skeleton. (Even if your project isn't Node-based, you'll need a `package.json`
to track the tooling used by **DevPail**.)

### Ports

**DevPail** uses ports 3000-3009 inside the container. 3000 is always a
BrowserSync server, and 3009 is the BrowserSync UI. Use of the remaining ports
is optional and application-specific. You specify which you want mapped onto
your local ports with the `ports` key, which is an array of "`host`:`container`"
mappings, i.e.:

```json
"devpail": {
    "ports": [ "8080:3000", "4044:3009" ]
}
```

This makes the container's browser-sync available as `localhost:8080`, and the
browser-sync UI as `localhost:4044`. **You MUST list the container's port 3000
FIRST!**

### Compiler and Server Plugins

In **DevPail**'s parlance, _Server_ and _Compiler_ plugins manage distinct parts
of the development and build processes.

During development, a _Server_ plugin can affect the BrowserSync configuration,
and it may start a long-running process (such as a Flask instance) which is
proxied (or not) into BrowserSync. A _Compiler_ plugin has access to the
BrowserSync instance, so it CANNOT alter the BrowserSync configuration.

During the `build` task, all the _Server_ plugins run in series (unless
otherwise configured), then all the _Compiler_ plugins run in parallel (unless
otherwise configured).

Plugins may not need specific configuration, in which case they can be listed by
name. If they need configuration, they are listed in an object which contains
that configuration. For example:

```JSON
"devpail": {
  "compilers": [
    "compiler/html/terser",
    {
      "plugin": "compiler/javascript/uglify",
      "source": "javascriptSource",
    },
    "compiler/css/clean-css"
  ],
  "Servers": [
    {
      "server": "server/flask",
      "environ": {
        "FLASK_CONFIG_VAR": "flask-config-value"
      }
    }
  ]
}
```

In this case, the `terser` and `clean-css` plugins will operate based on their
defaults, while the `uglify` and `flask` plugins will use the provided
configuration.

### Task plugins

Task plugins provide additional Gulp tasks which are added to the Gulp task list, thus:

```JSON
"devpail": {
  "tasks": [
    {
      "name": "serve",
      "plugin": "task/serve/serve-static"
    },
    {
      "name": [ "deploy", "deploy:setup" ],
      "plugin": "task/deploy/aws-s3",
      "bucket": "my-app-deploy-bucket"
    }
  ]
}
```

Each Task plugin is connected to one or more task names in the Gulp task
registry, as listed. Configuration parameters are as with other plugin types.

### Plugin Location

Each plugin is loaded either from the project source code, or via a CDN over the
wire.

To specify a local plugin, use the form `~dir/to/plugin` (note the leading `~`),
which will try to find `plugin.js` in `dir/to` in your source tree.

Otherwise, the plugin is loaded from the CDN defined by the `devpail.moduleCDN`
parameter (if any) in your `package.json`.

If you need to operate without internet access, **DevPail** can provide a "local
CDN" with these steps:

1. Run `devpail --cdn` from a terminal. The "local CDN" will run on port 8000 unless you specify a port number after `--cdn`.
1. Set `devpail.moduleCDN` to `http://host.docker.internal:8000`, or whatever port you specified above.
1. Run `devpail` as usual!

## MetaTasks

Some tasks are composed of other tasks; these are *metatasks*. For example, the default `build` task is implemented by running the `clean:build` task, followed by the `devpail:build` task.

You define your own *metatasks*, including **overriding** a built-in *metatask*, thusly:

```json
"devpail": {
  "metatasks": {
    "my-meta-task": [
      "task-1",
      [ "task-2", "task-3" ], 
      "task-4"
    ]
  }
}
```

This creates a new task, *my-meta-task*, which runs *task-1*, then runs *task-2*
and *task-3* simultaneously (in parallel), then runs *task-4*.

The default *metatasks* are:

* `build`: 'clean:build', 'devpail:build'
* `clean`: [ 'clean:build', 'clean:dev' ]



## Running the development environment

To run the environment, once configured:

```console
$ devpail
DevPail: Starting your development environment...
[19:57:29] Using gulpfile ~/app/gulpfile.js
[19:57:29] Starting 'default'...
[Browsersync] Access URLs:
 ----------------------------
 Local: http://localhost:3000
 ----------------------------
    UI: http://localhost:3009
 ----------------------------
[Browsersync] Serving files from: dev/
```

> The first time you run `devpail`, it will have to install at least a few modules, possibly many, which may take a bit of time. But after the first run, it'll go pretty quickly.

Please remember that the 3000 port specified by BrowserSync is **inside** the
container. The port you should browse to is defined in your `devpail.ports`
configuration.

You can get a shell into the **DevPail** container:

```console
$ devpail --shell
DevPail: Running a shell in your project container...
pn@your-project-f44be227:~/app$ 
```

From there you can:

* Install additional project dependencies
  * `npm install --save some-package-name`
  * `poetry add some-package-name`
* Run `gulp --tasks` or `gulp some-task-name`
* Most anything else you'd do from a command prompt.

You can also pass Gulp task name(s) and parameters to the Gulp instance via `devpail`:

```console
$ devpail --tasks
DevPail: Starting your development environment...
[20:05:32] Tasks for ~/app/gulpfile.js
[20:05:32] ├── build
[20:05:32] ├── clean
[20:05:32] ├── clean:dev
[20:05:32] ├── clean:deploy
[20:05:32] ├── default
[20:05:32] ├── deploy
[20:05:32] └── deploy:setup
```

## Setting Secrets via .env

**DevPail** will inject environment variables specified in an `.env` file at the
top level of your source tree into your container's environment.

## TODO: List of Plugins

## TODO: Developing Plugins

## How to Contribute

We gratefully accept:

* [Suggestions][gh-suggestions]
* [Bug Reports][gh-issues]
* [Pull Requests][gh-pulls]
* [Donations][bmac]

<!-- Please see [CONTRIBUTING.md](/CONTRIBUTING.md) for all the details. -->

## Donating

I release (most of?) my code under the MIT license, so it's free to use any way
you'd like. If you use it and appreciate the efforts I've put into it, please
consider [Buying Me A Coffee][bmac]!

<a href="https://www.buymeacoffee.com/tdesposito" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-violet.png" alt="Buy Me A Coffee" style="height: 48px !important;width: 174px !important;" ></a>

## License

Please see the [License](LICENSE).

<!-- Links  -->
[bmac]: https://www.buymeacoffee.com/tdesposito
[gh-issues]: https://github.com/tdesposito/devpail/issues
[gh-license]: https://github.com/tdesposito/devpail/blob/main/LICENSE
[gh-pulls]: https://github.com/tdesposito/devpail/pulls
[gh-repo]: https://github.com/tdesposito/devpail
[gh-suggestions]: https://github.com/tdesposito/devpail/discussions
[npm-package]: https://npmjs.org/package/devpail
[npm-versions]: https://www.npmjs.com/package/devpail?activeTab=versions
[pypi-project]: https://pypi.org/project/devpail/
[rtd-latest]: https://devpail.readthedocs.io/en/latest/
[shield-bmac]: https://img.shields.io/badge/Donate-Buy%20me%20a%20coffee-purple.svg
[shield-gh-latest-tag]: https://img.shields.io/github/v/tag/tdesposito/devpail
[shield-gh-json-version]: https://img.shields.io/github/package-json/v/tdesposito/devpail/main?label=latest-version
[shield-gh-license]: https://img.shields.io/github/license/tdesposito/devpail
[shield-npm-downloads-month]: https://img.shields.io/npm/dm/devpail
[shield-npm-version]: https://img.shields.io/npm/v/devpail
[shield-pypi-version]: https://img.shields.io/pypi/v/devpail
[shield-pypi-python-versions]: https://img.shields.io/pypi/pyversions/devpail
[shield-rtd]: https://readthedocs.org/projects/devpail/badge/?version=latest
[gh-license]: https://github.com/tdesposito/devpail/blob/main/LICENSE
[shield-gh-license]: https://img.shields.io/github/license/tdesposito/devpail

[Docker]: https://docker.com
[GulpJS]: https://gulpjs.com

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

* The format is based on [Keep a Changelog](https://keepachangelog.com/)
* This project adheres to [Semantic Versioning](https://semver.org/)
* This project uses [cvbump](https://github.com/tdesposito/ChangelogVersionTool) to maintain this changelog.

## [2.1.1](https://github.com/tdesposito/DevPail/releases/tag/v2.1.1) - 2022-02-18

### Fixed

* Fixed metatask injection ([e1f0b1c](https://github.com/tdesposito/DevPail/commit/e1f0b1c83345bd06fe254f94156134e5574804fb) by Todd Esposito)

## [2.1.0](https://github.com/tdesposito/DevPail/releases/tag/v2.1.0) - 2022-02-18

### Added

* Add pycache redirection ([7042399](https://github.com/tdesposito/DevPail/commit/704239975f206f5220b3b48a3deb24f429a76b5a) by Todd Esposito)

### Fixed

* Fixed compiler/css/sass to watch subdirs ([80e2c8f](https://github.com/tdesposito/DevPail/commit/80e2c8f0ca6e117e2788b4662e582f0361b1b814) by Todd Esposito)
* Fixed metatask config to replace rather than merge the defaults ([c82543c](https://github.com/tdesposito/DevPail/commit/c82543ccd49c6e768e8c4ed8f363d3497ec08357) by Todd Esposito)

### Other Updates

* Updated init dir list - `deploy` no longer needed ([3a81900](https://github.com/tdesposito/DevPail/commit/3a819003e250d2f76b8260d95273950fed7f4ca7) by Todd Esposito)

## [2.0.0](https://github.com/tdesposito/DevPail/releases/tag/v2.0.0) - 2022-01-27

### Added

* Added "cvbump" task for changelog and versioning ([ae21494](https://github.com/tdesposito/DevPail/commit/ae214947d309685c1bc1e369fca8d37b53cc55df) by Todd Esposito)
* Added .serverlessrc to the list of bundled secrets ([8df6160](https://github.com/tdesposito/DevPail/commit/8df61602a70167283ff44e903a74cf924c8dfbeb) by Todd Esposito)
* Added support for "metatasks" ([6009be2](https://github.com/tdesposito/DevPail/commit/6009be252cbfb951861573b58a7767db29463172) by Todd Esposito)
  <br>_- revised the inbuild "build" task_
  <br>_- created a default "build" metatask_
  <br>_- revised the inbuild "clean" task_
  <br>_- created a default "clean" metatask_
  <br>_- misc refactors_
* Added AWS Lambda plugin ([1870c91](https://github.com/tdesposito/DevPail/commit/1870c91fb02fc1e29f88b916bd0c2c3b1fe3fdb2) by Todd Esposito)
* Added more configuration options to AWS EB ([4f29b8c](https://github.com/tdesposito/DevPail/commit/4f29b8c4c25d4c0d8c9a9b2e9a7f6b1c824a763f) by Todd Esposito)
* Added support for simplified compiler/server specs ([c577b03](https://github.com/tdesposito/DevPail/commit/c577b03430c0a6633cd568e489cbcf62313d8e86) by Todd Esposito)
  <br>_You can now list compilers/servers without config as just the plugin name_
* Added cwd hash to project name ([77ce227](https://github.com/tdesposito/DevPail/commit/77ce227d31a7c76a8be87bd9d8090bee28934276) by Todd Esposito)
  <br>_This avoids volume-name collisions for projects sharing a folder name_
* Added load-balancer config support to aws-eb ([497589c](https://github.com/tdesposito/DevPail/commit/497589cd6f14f7759140298ebc55f8c973bb9481) by Todd Esposito)
* Added copy-files plugin ([26c9882](https://github.com/tdesposito/DevPail/commit/26c98821a0c700925c4e947e89c7d89bc533295b) by Todd Esposito)
* Added tests for react plugin ([0aca8b1](https://github.com/tdesposito/DevPail/commit/0aca8b13cf079e45b9c91930aa6ad9a81180625d) by Todd Esposito)
* Add to basic test for imagemin plugin ([36e6d39](https://github.com/tdesposito/DevPail/commit/36e6d39f0725f7945ddc6a3a21f7fb4a54f175bb) by Todd Esposito)
* Added 'serve-static' task to serve the as-built app ([319da0f](https://github.com/tdesposito/DevPail/commit/319da0fc96c36447c1d38b11ce23f9aecaa42931) by Todd Esposito)
* Added aws-eb deployment task ([87caf22](https://github.com/tdesposito/DevPail/commit/87caf22b8a48d96bbc610532e96767f5a7414e95) by Todd Esposito)
* Added s3 setup feature. ([599d024](https://github.com/tdesposito/DevPail/commit/599d0240fd0919da977a7b08ef307966f51ee56b) by Todd Esposito)
  <br>_Naming and executing a task named like "*:setup" will create your hosting bucket._
* Added cli --cdn option ([93971b7](https://github.com/tdesposito/DevPail/commit/93971b7fc6c49eb7ce88013ea10d7fe3714125cd) by Todd Esposito)
  <br>_Running `devpail --cdn` now serves the locally installed plugins._
* Add "imagemin" compiler ([aac09bd](https://github.com/tdesposito/DevPail/commit/aac09bd3b532ac7c18a761b662de869d5e7836cd) by Todd Esposito)
* Add version/help options to the cli ([a19545e](https://github.com/tdesposito/DevPail/commit/a19545ef1e7c77fbd07c664ed275b0e67f4dd337) by Todd Esposito)
* Add aws-s3 deploy task ([a9f66ce](https://github.com/tdesposito/DevPail/commit/a9f66ced744233bab1ee9bb1363265f1102e1f5e) by Todd Esposito)
* Add .env injection ([cc890bc](https://github.com/tdesposito/DevPail/commit/cc890bcbde5c885917966d3ec64abf12da823dd4) by Todd Esposito)
* Added "lastRun" optimizations ([c71fa89](https://github.com/tdesposito/DevPail/commit/c71fa89044f39b7294d9396ade469ce8bb1f8e7f) by Todd Esposito)
* Add dependancy handling and building ([40f76fd](https://github.com/tdesposito/DevPail/commit/40f76fd99a5526c008f27858a3cd469707aa95b8) by Todd Esposito)
  <br>_- remember which dependancies have already been installed_
  <br>_- add the `build` task_
  <br>_- update the build tasks for the compilers_

### Fixed

* Fixed webpack-missing-process/path ([29bea41](https://github.com/tdesposito/DevPail/commit/29bea41c5daa233ecb85510153aceac819b75d32) by Todd Esposito)
* Fixed proxy and react plugins to support hybrids ([1596c32](https://github.com/tdesposito/DevPail/commit/1596c324544f4ef60a2aa3102d6dd9a6b9ed2768) by Todd Esposito)
* Fix --cdn to listen on 0.0.0.0 ([75d6921](https://github.com/tdesposito/DevPail/commit/75d69215f3cfe0ceb949affc374017d8b61d676f) by Todd Esposito)
* Fixed flask plugin (multiple fixes) ([cb69e4b](https://github.com/tdesposito/DevPail/commit/cb69e4b4a1a114aa51e83c4219d1adcad62e4638) by Todd Esposito)

src/cli.js
#!/usr/bin/env node

// This file is part of DevPail, (https://github.com/tdesposito/DevPail)
// Copyright (C) Todd D. Esposito 2021.
// Distributed under the MIT License (see https://opensource.org/licenses/MIT).

const { spawnSync, execSync } = require('child_process')
const fs = require('fs-extra')
const os = require('os')
const path = require('path')

const status = processOptions()
if (status) {
    console.log(`\nDevPail: ${status}\n`)
}

function buildImage(opts) {
    console.log(`DevPail: Building image "devpail:${opts.tag}"...`)

    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'devpail-'))
    const fromDir = os.homedir()
    const toDir = `${tmpDir}/homedir`

    fs.copySync(path.resolve(__dirname, '..', 'imagesrc'), tmpDir)

    if (fs.existsSync(`${fromDir}/.ssh`)) {
        fs.copySync(`${fromDir}/.ssh`, `${toDir}/.ssh`, { dereference: true })
    }

    for (var file of ['config', 'credentials']) {
        if (fs.existsSync(`${fromDir}/.aws/${file}`)) {
            fs.ensureDirSync(`${toDir}/.aws`)
            fs.copySync(`${fromDir}/.aws/${file}`, `${toDir}/.aws/${file}`, { dereference: true })
        }
    }

    for (var file of ['.gitconfig', '.git-credentials', '.serverlessrc']) {
        if (fs.existsSync(`${fromDir}/${file}`)) {
            fs.copySync(`${fromDir}/${file}`, `${toDir}/${file}`, { dereference: true })
        }
    }


    spawnopts = {
        cwd: tmpDir,
        shell: true,
        stdio: 'inherit'
    }

    spawnSync(
        "docker",
        [
            "build",
            "-t",
            `devpail:${opts.tag}`,
            ...opts.vars.flatMap(v => ['--build-arg', v]),
            "."
        ],
        spawnopts
    )
    console.log("DevPail: Cleaning up...")
    spawnSync("docker", "image prune --force".split(' '), spawnopts)

    fs.rmSync(tmpDir, { recursive: true })

    return `WARNING: Do NOT publish or share "devpail:${opts.tag}"\n\tIt contains YOUR SECRETS, possibly including ssh keys, aws tokens, and git credentials!`
}


function cleanProject() {
    console.log('DevPail: cleaning up artifacts...')
    var project = createProjectName()
    spawnSync(
        "docker",
        [
            "volume",
            "rm",
            `${project}-DevPail-Tooling`
        ],
        {
            shell: true,
            stdio: 'inherit'
        }
    )
    return null
}


function createProjectName() {
    return path.basename(process.cwd()) 
        + '-' 
        + require('crypto')
            .createHash('md5')
            .update(process.cwd())
            .digest('hex')
            .slice(0, 8)
}


function initProject() {
    console.log('DevPail: Setting defaults for your project...')
    var pkg = JSON.parse(fs.readFileSync('./package.json').toString())
    if (pkg.devpail === undefined) {
        pkg.devpail = {}
    }
    ['compilers', 'ports', 'servers', 'tasks'].forEach((key) => {
        if (pkg.devpail[key] === undefined) {
            pkg.devpail[key] = []
        }
    })
    if (pkg.devpail.ports.length === 0) {
        pkg.devpail.ports = ['3000:3000']
    }
    if (pkg.devpail.imageTag === undefined) {
        pkg.devpail.imageTag = 'default'
    }
    fs.writeFileSync('./package.json', JSON.stringify(pkg, null, 2))
}


function killProject() {
    var project = createProjectName()
    var id = execSync(`docker container ls --filter "name=${project}-DevPail" --format "{{.ID}}"`).toString()
    if (id) {
        console.log(`DevPail: killing active container for ${project}...`)
        spawnSync(
            'docker',
            [
                "container",
                "kill",
                id
            ],
            {
                shell: true,
                stdio: 'inherit'
            }
        )
    } else {
        console.log(`DevPail: no active container for ${project}...`)
    }
}


function processOptions() {
    var args = process.argv.slice(2)
    var opts = { tag: 'default', commands: [], params: [] }
    if (args && args[0] === '--version') {
        return 'version 1.1.0'
    } else if (args && args[0] === '--help') {
        console.log("usage: devpail [options] [gulp option(s)/task(s)...]\n")
        console.log("Options:")
        console.log("\t--build [tag] [args...] - build a new DevPail docker image")
        console.log("\t--cdn [port] - serve the included Gulp plugins locally")
        console.log("\t--clean - remove the project's docker volume")
        console.log("\t--init - add/update a 'devpail' key in the local package.json")
        console.log("\t--kill - kill the running DevPail environment")
        console.log("\t--shell - shell into the DevPail environment")
        console.log("\n")
        console.log("With none of the above specified, everything on the command\nline will be arguments to the Gulp process.")
        console.log("\n")
    } else if (args && args[0] === '--build') {
        args.shift()
        if (args.length) {
            opts.tag = args.shift()
        }
        opts.vars = args
        return buildImage(opts)
    } else if (args && args[0] === '--cdn') {
        args.shift()
        serveCDN(args)
    } else if (args && args[0] === '--init') {
        return initProject()
    } else if (args && args[0] === '--clean') {
        return cleanProject()
    } else if (args && args[0] === '--kill') {
        return killProject()
    } else {
        if (args && args[0] === '--shell') {
            console.log("DevPail: Running a shell in your project container...")
            opts.params = ['--entrypoint', '/bin/bash']
            args.shift()
            opts.commands = args
        } else {
            console.log("DevPail: Starting your development environment...")
            opts.commands = args
        }
        return runContainer(opts)
    }
}


function runContainer(opts) {
    var tag, ports
    var project = createProjectName()
    try {
        pkg = require(`${process.cwd()}/package.json`)
        tag = pkg.devpail?.imageTag || 'default'
        ports = pkg.devpail?.ports || ['3000:3000']
    } catch (error) {
        return "DevPail: I couldn't find or parse your 'package.json'."
    }

    opts.params.push(...['--name', `${project}-DevPail-${(new Date().getTime()).toString().slice(6)}`])
    opts.params.push(...['--hostname', project])
    opts.params.push(...['-v', `${project}-DevPail-Tooling:/home/pn/app`])
    opts.params.push(...['-v', `${process.cwd()}:/home/pn/app/src`])
    if (fs.existsSync(`${process.cwd()}/.env`)) {
        console.log('DevPail: Injecting your .env variables')
        opts.params.push(...['--env-file', `${process.cwd()}/.env`])
    }
    opts.params.push(...['-e', `DEVPAIL_BS_PORT=${ports[0].split(':')[0]}`])
    opts.params.push(...ports.flatMap(p => ['-p', p]))
    
    spawnSync(
        "docker",
        [
            "run",
            "-it",
            "--rm",
            ...opts.params,
            `devpail:${tag}`,
            ...opts.commands
        ],
        {
            shell: true,
            stdio: 'inherit'
        }
    )
    return null
}


function serveCDN(opts) {
    var port = opts[0] || "8000"
    console.log(`DevPail: serving plugins on host.docker.internal:${port}...`)
    spawnSync(
        "npx",
        [
            "--quiet",
            "serve",
            "-l",
            `tcp://0.0.0.0:${port}`
        ],
        {
            cwd: path.resolve(__dirname, '..', 'gulp'),
            shell: true,
            stdio: 'inherit',
        }
    )
}